include/forward_index_builder.hpp

```diff
@@ -0,0 +1,288 @@
+            std::string term;
+            std::ptrdiff_t batch_term_id = 0;
+            while (std::getline(terms_is, term)) {
+                term_map[term][batch] = batch_term_id++;
+            }
+            id_mappings[batch].resize(batch_term_id);
+        }
+
+        logger() << "Mapping IDs and writing terms\n";
+        std::ptrdiff_t term_id = 0;
+        for (auto const &[term, idmap] : term_map) {
+            term_os << term << '\n';
+            for (auto const &[batch, batch_term_id] : idmap) {
+                id_mappings[batch][batch_term_id] = term_id;
+            }
+            ++term_id;
+        }
+
+        logger() << "Remapping IDs\n";
+        for (auto batch : enumerate(batch_count)) {
+            auto &mapping = id_mappings[batch];
+            writable_binary_collection coll(batch_file(basename, batch).c_str());
+            for (auto doc_iter = ++coll.begin(); doc_iter != coll.end(); ++doc_iter) {
+                for (auto &term_id : *doc_iter) {
+                    term_id = mapping[term_id];
+                }
+            }
+        }
+
+        logger() << "Concatenating batches\n";
+        std::ofstream os(basename);
+        write_header(os, document_count);
+        for (auto batch : enumerate(batch_count)) {
+            std::ifstream is(batch_file(basename, batch));
+            is.ignore(8);
+            os << is.rdbuf();
+        }
+
+        logger() << "Done.\n";
+    }
+
+    void build(std::istream &             is,
+               std::string const &        output_file,
+               read_record_function_type  next_record,
+               process_term_function_type process_term,
+               std::ptrdiff_t             batch_size,
+               std::size_t                threads)
+    {
+        Document_Id    first_document{0};
+        std::ptrdiff_t batch_number = 0;
+
+        std::vector<Record> record_batch;
+        tbb::task_group     batch_group;
+        tbb::concurrent_bounded_queue<int> queue;
+        queue.set_capacity((threads - 1) * 2);
+        while (true) {
+            std::optional<Record> record = std::nullopt;
+            try {
+                if (not (record = next_record(is))) {
+                    auto last_batch_size = record_batch.size();
+                    Batch_Process bp{
+                        batch_number, std::move(record_batch), first_document, output_file};
+                    queue.push(0);
+                    batch_group.run([bp = std::move(bp), process_term, this, &queue]() {
+                        run(std::move(bp), process_term);
+                        int x;
+                        queue.try_pop(x);
+                    });
+                    ++batch_number;
+                    first_document += last_batch_size;
+                    break;
+                }
+            } catch (Warc_Format_Error &err) {
+                continue;
+            }
+            if (record->type() != "response") {
```

Comment:
Not too generic. Once we plug another parse might not be `response` anymore.
How about a `is_valid()` function?